### ISSUE TITLE
C-API: add a function to retrieve page content stream

### ISSUE THREAD

**Issue by ancwrd1** [qpdf/qpdf#596]
In order to support several use cases for the manipulation of PDF pages it would be great to add C functions which can retrieve and set page contents.
Examples: N-up document formatting, booklet formatting, mirroring, rotating, etc. In these cases the original pages are transformed into a dictionary of xobjects and imposed onto an empty page (e.g. 2x2, etc) and their contents are copied verbatim with eventual matrix transformations applied.

I was thinking about higher-level C function which can create a stream (via `QPDFObjectHandle::newStream`) and write page contents into it.

```c
qpdf_oh qpdf_get_page_content_stream(qpdf_data data, qpdf_oh page);
```

It would call `page.getPageContents()` and construct an instance of StreamDataProvider-inherited class with the returned list. This class will concatenate the contents (code shamelessly taken from pdftopdf cups filter, all credits go there):

```c++
void CombineFromContents_Provider::provideStreamData(int, int, Pipeline *pipeline) {
    Pl_Concatenate concat("concat", pipeline);
    for (auto &content: m_contents) {
        content.pipeStreamData(&concat, true, false, false);
    }
    concat.manualFinish();
}
```

The C function will then construct a new stream via `QPDFObjectHandle::newStream` and call `stream.replaceStreamData` passing over a pointer to the StreamDataProvider instance.

So the returned stream can be placed into a dictionary as a xobject.

**Comment by jberkenbilt** [comment]
I'm thinking a more likely approach to this would be to allow retrieval of raw and filtered stream data and replacement of stream data using caller-managed memory. Something like

```
QPDF_ERROR_CODE qpdf_oh_get_stream_data(qpdf_data qpdf, qpdf_oh oh, QPDF_BOOL filtered, char** bufp, size_t* len);
qpdf_oh qpdf_oh_new_stream(qpdf_data qpdf);
qpdf_oh qpdf_oh_get_dict(qpdf_data qpdf, qpdf_oh stream_oh);
void qpdf_oh_replace_stream_data(qpdf_data qpdf, char const* buf, size_t len);
```

If that would meet your needs, I think I could work it in for 10.5, which I'd like to get out in the next few days.

**Comment by jberkenbilt** [comment]
I could probably add a `qpdf_oh_get_page_content_data` that would take the same arguments as `get_stream_data` but would take care of the concatenation as well.

All that said, nup requires a little more logic regarding form XObjects and so forth. I added code to the library to help with that after pdftopdf was written and actually did so in consultation with the people who worked on that code. I was heavily involved in that effort but did not write the actual code. Anyway current qpdf has library functions to do some of what that code is doing, but I don't have any interest in replicating the page helper code and the rest of the library in the C API. I think the functions above would round out the existing interface, perhaps along with copyForeignObject. That's probably as far as I would go. So add this:

```
qpdf_oh qpdf_oh_copy_foreign_object(qpdf_data qpdf, qpdf_data other_qpdf, qpdf_oh foreign_oh);
QPDF_ERROR_CODE qpdf_oh_get_page_content_data(qpdf_data qpdf, qpdf_oh page_oh, char** bufp, size_t* len);
```

**Comment by jberkenbilt** [comment]
The buffers will be `unsigned char *` for consistency with the C++ API.

**Comment by jberkenbilt** [comment]
Here's what I've got. Could still change. I'll see if I can build it tomorrow. Should be pretty simple.

```C
    /* Note about foreign objects: the C API does not have enough
     * information in the value of a qpdf_oh to know what QPDF object
     * it belongs to. To uniquely specify a qpdf object handle from a
     * specific qpdf_data instance, you always pair the qpdf_oh with
     * the correct qpdf_data. Otherwise, you are likely to get
     * completely the wrong object if you are not lucky enough to get
     * an error about the object being invalid.
     */

    /* Copy foreign object: the qpdf_oh returned belongs to `qpdf`,
     * while `foreign_oh` belongs to `other_qpdf`.
     */
    QPDF_DLL
    qpdf_oh qpdf_oh_copy_foreign_object(
        qpdf_data qpdf, qpdf_data other_qpdf, qpdf_oh foreign_oh);

    /* STREAM FUNCTIONS */

    /* These functions provide basic access to streams and stream
     * data. They are not as comprehensive as what is in
     * QPDFObjectHandle, but they do allow for working with streams
     * and stream data as caller-managed memory.
     */

    /* Create a new stream. Use qpdf_oh_get_dict to get (and
     * subsequently modify) the stream dictionary if needed. See
     * comments in QPDFObjectHandle.hh for newStream() for additional
     * notes. You must call qpdf_oh_replace_stream_data to provide
     * data for the stream.
    */
    QPDF_DLL
    qpdf_oh qpdf_oh_new_stream(qpdf_data qpdf);

    /* Get stream data as a buffer. The buffer is allocated with
     * malloc and must be freed by the caller. The size of the buffer
     * is stored in *len.
     */
    QPDF_DLL
    QPDF_ERROR_CODE qpdf_oh_get_stream_data(
        qpdf_data qpdf, qpdf_oh stream_oh, QPDF_BOOL filtered,
        unsigned char** bufp, size_t* len);

    /* This function returns the concatenation of all of a page's
     * content streams as a single, dynamically allocated buffer. As
     * with qpdf_oh_get_stream_data, the buffer is allocated with
     * malloc and must be freed by the caller.
     */
    QPDF_DLL
    QPDF_ERROR_CODE qpdf_oh_get_page_content_data(
        qpdf_data qpdf, qpdf_oh page_oh,
        unsigned char** bufp, size_t* len);

    /* The data pointed to by bufp will be copied by the library. It
     * does not need to remain valid after the call returns.
     */
    QPDF_DLL
    void qpdf_oh_replace_stream_data(
        qpdf_data qpdf, unsigned char const* buf, size_t len);
```

**Comment by ancwrd1** [comment]
Looks nice, gives more functionality that originally requested.

**Comment by jberkenbilt** [comment]
I ended up with something slightly different, but once it gets through CI, it will be merged, and it will be in for 10.5.
